src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,166 @@
+                head = newNode;
+            } else {
+                current.prev.next = newNode;
+            }
+            current.prev = newNode;
+            size++;
+        }
     }
 
     @Override
-    public void addAll(List<T> list) {
+    public boolean addAll(List<T> list) {
+        if (list == null || list.isEmpty()) {
+            return false;
+        }
+        for (T element : list) {
+            add(element);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return findNodeByIndex(index).data;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> targetNode = findNodeByIndex(index);
+        T oldValue = targetNode.data;
+        targetNode.data = value;
+        return oldValue;
     }
 
+
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> toRemove = findNodeByIndex(index);
+        return unlink(toRemove);
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        while (current != null) {
+            if ((object == null && current.data == null)
+                    || (object != null && object.equals(current.data))) {
+                unlink(current);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
-    @Override
-    public int size() {
-        return 0;
+    private void checkIndexForAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException(
+                    "Index outside the list: " + index);
+        }
     }
 
-    @Override
-    public boolean isEmpty() {
-        return false;
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException(
+                    "Index outside the list: " + index);
+        }
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        Node<T> current;
+        if (index < size / 2) {
+            current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+        } else {
+            current = tail;
+            for (int i = size - 1; i > index; i--) {
+                current = current.prev;
+            }
+        }
+        return current;
+    }
+
+    private T unlink(Node<T> node) {
+        T element = node.data;
+
+        if (node.prev == null) { // Видаляємо головний вузол
+            head = node.next;
+        } else {
+            node.prev.next = node.next;
+        }
+
+        if (node.next == null) { // Видаляємо останній вузол
+            tail = node.prev;
+        } else {
+            node.next.prev = node.prev;
+        }
+
+        size--;
+        return element;
+    }
+
+    private static class Node<T> {
```

Comment:
Don't use getters and setters in class Node. We have access to private fields of the inner class, so using getters or setters is redundant.